### PR TITLE
First draft of app

### PR DESCRIPTION
**Pull request by justinvdm**

### CODE REVIEW

**Comment by justinvdm** [comment]
Ready for review. The big diff size is mostly because of the builds and fixtures, the actual code isn't that much.

File: src/app.js [hunk 1]
```diff
@@ -1,22 +1,234 @@
...
+        self.submit_report = function(content) {
+            return self
+            .ureporter.reports.submit(content)
+            .then(function(result) {
+                if (result.accepted) {
+                    return result.response;
+                }
+                else {
+                    // TODO not sure what to do here?
+                    return;
```
Comment (hodgestar) [inline src/app.js]: Maybe lets at least return something to display, e.g. "Report rejected. Please try again."?

File: src/app.js [hunk 2]
```diff
@@ -1,22 +1,234 @@
...
+            });
+        });
+
+        // TODO what to put as default question?
+        self.states.add('states:results:choose', function(name) {
+            return self.ureporter.polls.topics().then(function(topics) {
+                return new ChoiceState(name, {
+                    question: (
+                        "Choose poll:"),
+                        choices: topics.map(function(topic) {
```
Comment (hodgestar) [inline src/app.js]: Indentation here seems a little odd.

File: src/app.js [hunk 3]
```diff
@@ -1,22 +1,234 @@
...
+                return new EndState(name, {
+                    text: self.format_summary(summary),
+                    next: 'states:start'
+                });
+            });
+        });
+
+        self.states.add('states:reports:submit', function(name) {
+            return new FreeText(name, {
+                question: "Enter message:",
```
Comment (hodgestar) [inline src/app.js:235]: I think `Enter report:` might make more sense here.

File: src/app.js (hunk 1 above)
Comment (hodgestar) [inline src/app.js]: I see we do that later on.

**Comment by hodgestar** [comment]
Other than two minor quibbles, looks good.

File: src/app.js (hunk 3 above)
Comment (justinvdm) [inline src/app.js:235]: This is the text we were given in the one doc, but it is a bit meh :) I've got a feeling they will be changing soon, not sure if it is worth changing.

File: src/app.js (hunk 2 above)
Comment (justinvdm) [inline src/app.js]: Really not sure what I was trying to do here :) thanks

File: src/app.js (hunk 3 above)
Comment (hodgestar) [inline src/app.js:235]: If it was in the doc, let's keep it for now.

**Comment by justinvdm** [comment]
Ready for re-review.

**Comment by hodgestar** [comment]
:+1:

I think it's possible to reduce the complexity of some of the states (specifically those involving rejected responses to Ureport) by using `opts.error` from http://vumi-jssandbox-toolkit.readthedocs.org/en/latest/states.html#FreeText but I don't want to hold up this PR for that. If it looks like a good idea, let's do a separate PR for it later.

**Comment by justinvdm** [comment]
@hodgestar I avoided `opts.error` since we only know the error response later on. Maybe `FreeText` should grow a `set_error` or something? Not sure if its okay to just tack it on (`state.in_error = response`).

**Comment by justinvdm** [comment]
@hodgestar How about we change `opts.check` so that it make it return a a string, and remove `opts.check`. If the return value is not a string, we take it as valid user input. Backbone does something similar for `Model.validate()`, so maybe js people would be a little familiar with it?

**Comment by justinvdm** [comment]
@hodgestar Ah, so [`Model.validate()`](http://backbonejs.org/#Model-validate) takes errors objects and strings as invalid user input. What are your thoughts on something like that?

**Comment by hodgestar** [comment]
I like the `opts.check` idea -- I've wanted to be able to customize the error string for `FreeText` states for awhile. Maybe `.set_error` isn't a bad idea either since it explicitly allows control of the error externally which will allow library users to do interesting things if they need to. Definitely all for a separate PR though.

**Pull request merged**